xds/internal/xdsclient/authority.go

```diff
@@ -0,0 +1,229 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xdsclient
+
+import (
+	"errors"
+	"fmt"
+
+	"google.golang.org/grpc/xds/internal/xdsclient/bootstrap"
+	"google.golang.org/grpc/xds/internal/xdsclient/load"
+	"google.golang.org/grpc/xds/internal/xdsclient/pubsub"
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource"
+)
+
+const federationScheme = "xdstp"
+
+// findAuthority returns the authority for this name. If it doesn't already
+// exist, one will be created.
+//
+// Note that this doesn't always create new authority. authorities with the same
+// config but different names are shared.
+//
+// Caller must not hold c.authorityMu.
+func (c *clientImpl) findAuthority(n *xdsresource.Name) (retA *authority, _ error) {
+	scheme, authority := n.Scheme, n.Authority
+
+	c.authorityMu.Lock()
+	defer c.authorityMu.Unlock()
+	if c.done.HasFired() {
+		return nil, errors.New("the xds-client is closed")
+	}
+
+	defer func() {
+		// All returned authority from this function will be used by a watch,
+		// holding the ref here.
+		//
+		// Note that this must be done while c.authorityMu is held, to avoid the
+		// race that an authority is returned, but before the watch starts, the
+		// old last watch is canceled (in another goroutine), causing this
+		// authority to be removed, and then a watch will start on a removed
+		// authority.
+		//
+		// unref() will be done when the watch is canceled.
+		if retA != nil {
+			retA.ref()
+		}
```

Comment:
Why is this defer required? Or in other words, why would the following not work at the end of the function?

```golang
a, err := c.newAuthority(config)
if err != nil {
	return nil, fmt.Errorf("xds: failed to connect to the control plane for authority %q: %v", authority, err)
}
a.ref()
return a, nil
```

Reply:
Moved.